src/crimson/admin/osd_admin.cc

```diff
@@ -0,0 +1,258 @@
+  TestThrowHook  osd_test_throw_hook;
+
+ public:
+  OsdAdminImp(OSD* osd, CephContext* cct, crimson::common::ConfigProxy& conf)
+      : m_osd{ osd }
+      , m_cct{ cct }
+      , m_conf{ conf }
+      , osd_status_hook{ *this }
+      , send_beacon_hook{ *this }
+      , osd_test_throw_hook{ *this }
+  {}
+
+  ~OsdAdminImp()
+  {
+    // our registration with the admin_socket server was already removed by
+    // 'OsdAdmin' - our 'pimpl' owner. Thus no need for:
+    //   unregister_admin_commands();
+  }
+
+  seastar::future<> register_admin_commands()
+  {
+    static const std::vector<AsokServiceDef> hooks_tbl{
+      // clang-format off
+        AsokServiceDef{"status",      "status",        &osd_status_hook,
+                      "OSD status"}
+      , AsokServiceDef{"send_beacon", "send_beacon",   &send_beacon_hook,
+                      "send OSD beacon to mon immediately"}
+      , AsokServiceDef{"throw",       "throw",         &osd_test_throw_hook,
+                      ""}  // dev tool
+      , AsokServiceDef{"stoposd",     "stoposd",       &osd_test_throw_hook,
+                      ""}  // dev tool
+      , AsokServiceDef{"fthrow",      "fthrow",        &osd_test_throw_hook,
+                      ""}  // dev tool
+      // clang-format on
+    };
+
+    return m_cct->get_admin_socket()
+      ->register_server(AdminSocket::hook_server_tag{ this }, hooks_tbl)
+      .then([this](AsokRegistrationRes rr) { m_socket_server = rr; });
+  }
+
+  seastar::future<> unregister_admin_commands()
+  {
+    if (!m_socket_server.has_value()) {
+      logger().warn("{}: OSD asok APIs already removed", __func__);
+      return seastar::now();
+    }
+
+    AdminSocketRef srv{ std::move(m_socket_server.value()) };
+
+    auto admin_if = m_cct->get_admin_socket();
+    if (admin_if) {
+      return admin_if->unregister_server(AdminSocket::hook_server_tag{ this },
+                                         std::move(srv));
+    } else {
+      return seastar::now();
+    }
+  }
+};
+
+//
+//  some PIMPL details:
+//
+OsdAdmin::OsdAdmin(OSD*                          osd,
+                   CephContext*                  cct,
+                   crimson::common::ConfigProxy& conf)
+    : m_imp{ std::make_unique<crimson::admin::OsdAdminImp>(osd, cct, conf) }
+{}
+
+seastar::future<> OsdAdmin::register_admin_commands()
+{
+  return m_imp->register_admin_commands();
+}
+
+seastar::future<> OsdAdmin::unregister_admin_commands()
+{
+  if (m_imp)
+    return m_imp->unregister_admin_commands();
+  else
+    return seastar::now();
+}
+
+OsdAdmin::~OsdAdmin()
+{
+  // relinquish control over the actual implementation object, as that one
+  // should only be destructed after the de-registration
+  std::ignore = seastar::do_with(std::move(m_imp), [](auto&& imp) {
+    // test using sleep(). Change from 1ms to 1s:
+    return seastar::sleep(1ms).then(
+      [imp_ptr = imp.get()]() { return imp_ptr->unregister_admin_commands(); });
+  });
+}
```

Comment:
I don't really understand why we're doing this pimpl thing here.  unregister_admin_commands must have already been called in the event of an even remotely orderly shutdown (at_exit ensures that OSD::stop will have been called prior to ~OSD).  Why not just eliminate the pimpl thing and assert that unregister_admin_commands has been called?

Reply:
If we do choose do do it anyway, I don't want to leave seastar::sleep in production code.  If we need it for unit tests, we'll have to find another way of injecting those delays.

Reply:
Agree on the sleep. Removed.

Re the pimpl: its general use here is important, as otherwise I drag a lot of internals into osd.h.
And as I already have the structure - it makes sense to use it as a way to simplify the reasoning at destruction.